File: Easy/Reverse_LinkedList.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * }; 
 */

#include <bits/stdc++.h>
#include <iostream>
#include <stack>
using namespace std;

class Solution {
public:
    ListNode* reverseList(ListNode* head) {
        if(head == NULL){
            return 0;
        }
        
        stack<ListNode*> s;
        ListNode* node = head;
        
        
        while(node->next != NULL){
            s.push(node);
            node = node->next;
        }
        
        ListNode* ans = node;
        
        while(!s.empty()){
            node->next = s.top();
            s.pop();
            node = node->next;
        }
        node->next = NULL;
    
    
        return ans;
    }
}; 
 
/*
Notes:
could also use recursion for less memory usage

*/

File: Easy/Reverse_Linked_List.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    # First attempt: iterate through the linked list and add each element to a list, could also do a stack to be easier
    # then go through that list in reverse order or go through the stack popping each element and adding them to a new ListNode instance
    # runtime of O(n) and space of O(n) for the list/stack
    def reverseList(self, head: Optional[ListNode]) -> Optional[ListNode]:
        if not head:
            return head
        tempList = []
        while head:
            tempList.append(head.val)
            head = head.next

        newHead = ListNode(tempList[len(tempList)-1])
        tempHead = newHead
        for num in tempList[::-1][1:]:
            temp = ListNode(num)
            tempHead.next = temp
            tempHead = tempHead.next
        
        return newHead

        
        